File: service/src/main/java/com/loudsight/utilities/service/dispatcher/Dispatcher.java
package com.loudsight.utilities.service.dispatcher;

import com.loudsight.useful.entity.permission.Subject;
import com.loudsight.utilities.service.dispatcher.bridge.BridgeMessageType;

public interface Dispatcher {
        String BRIDGE_RETURN = "bridgeMessage";
    

    interface Subscription<Q, A> {
        long getId();

        A onEvent(Address to,
                  Address replyTo,
                  Subject recipient,
                  Subject sender,
                  Object payload,
                  BridgeMessageType publicationType);
//               

        void unsubscribe();

        boolean isActive();

        boolean isBridged();
    }

    /**
     * Subscribe to a topic, provide a handler that accepts
     */


    <Q, A> Subscription<Q, A> subscribe(Address to, MessageHandler<A> handler);

    <T> Subscription<Publication,T> bridge(Address to, MessageHandler<T> handler);

    <T> void publishAsync(Address to,
                      Subject recipient,
                      Subject publisher,
                      Object payload,
                      MessageHandler<T> handler);
    void publish(Address to,
                 Subject recipient,
                 Subject publisher,
                 Object payload,
                 BridgeMessageType publicationType);

    default void publish(Address to, Subject recipient, Subject publisher, Object payload) {
        publish(to, recipient, publisher, payload, BridgeMessageType.DIRECT);
    }
}

File: service/src/main/java/com/loudsight/utilities/service/dispatcher/Dispatcher.kt
package com.loudsight.utilities.service.dispatcher

import com.loudsight.meta.MetaRepository
import com.loudsight.utilities.permission.Subject
import com.loudsight.utilities.service.dispatcher.bridge.BridgeMessageType
import com.loudsight.utilities.service.dispatcher.bridge.BridgeMessageTypeMeta

interface Dispatcher {
    companion object {
        const val BRIDGE_RETURN = "bridgeMessage"
    }

    interface Subscription<Q, A> {
        fun getId(): Long

        fun onEvent(
            to: Address,
            replyTo: Address,
            recipient: Subject,
            sender: Subject,
            payload: Any?,
            publicationType: BridgeMessageType
        ): A?

        fun unsubscribe()

        fun isActive(): Boolean

        fun isBridged(): Boolean
    }

    /**
     * Subscribe to a topic, provide a handler that accepts
     */
    fun <Q, A> subscribe(to: Address, handler: MessageHandler<A>): Subscription<Q, A>

    fun bridge(to: Address, handler: MessageHandler<Unit>): Subscription<Publication, Unit>

    fun publishAsync(to: Address, recipient: Subject, publisher: Subject, payload: Any?, handler: MessageHandler<Unit>)

    fun publish(to: Address, recipient: Subject, publisher: Subject, payload: Any?, publicationType: BridgeMessageType)

    fun publish(to: Address, recipient: Subject, publisher: Subject, payload: Any?) {
        publish(to, recipient, publisher, payload, BridgeMessageType.DIRECT)
    }
}
